synapse/handlers/sync.py

```diff
@@ -1323,9 +1323,13 @@ async def generate_sync_result(
             # See https://github.com/matrix-org/matrix-doc/issues/1144
             raise NotImplementedError()
 
+        # If we have no since token (init sync), ensure any cached rooms for the user
+        # is first invalidated to avoid race conditions with invalidation-over-replication.
+        if not since_token:
+            self.store.get_rooms_for_user.invalidate((user_id,))
```

Comment:
I'm *really* hesitant to do this without fully understanding where the inconsistencies are coming in, as per https://github.com/matrix-org/synapse/issues/14154#issuecomment-1282305045. In particular, it's not obvious to me that invalidating just this cache won't lead to really weird inconsistencies.